File: 02.2-Clase_arreglo_trabajadores/cpp/clase_nomina.cpp
#include "clase_nomina.h"

Nomina::Nomina(string rutaArchivo, string nombreCompania) {
    this->nombreCompania = nombreCompania;
    this->rutaArchivo = rutaArchivo;
    int contador = 0;
    string line, word, fileContent = "";

    fstream file(rutaArchivo, ios::in);
    if(file.is_open()) {

        getline(file, line);
        while(getline(file, line)) {
            fileContent += line + "\n";
            contador++;
        }
        file.close();

        this->empleados = new Arreglo<Trabajador>(contador);

        contador = 0;
        stringstream ss(fileContent);
        while (getline(ss, line, '\n')) {
            stringstream ss2(line);
            vector<string> datos;
            while (getline(ss2, word, ',')) {
                datos.push_back(word);
            }

            Trabajador trabajadorTemp;
            trabajadorTemp.setNumeroTrabajador(stoi(datos[0]));
            trabajadorTemp.setNombre(datos[1]);
            trabajadorTemp.setPaterno(datos[2]);
            trabajadorTemp.setMaterno(datos[3]);
            trabajadorTemp.setHorasExtra(stoi(datos[4]));
            trabajadorTemp.setSueldoBase(stod(datos[5]));
            trabajadorTemp.setAnioIngreso(stoi(datos[6]));

            this->empleados->setItem(contador, trabajadorTemp);
            contador++;
        }
    } else {
        cout<<"No se pudo abrir el archivo\n";
    }
}

Nomina::~Nomina()  {
    delete this->empleados;
}

void Nomina::calcularSueldo() {
    for(int i = 0; i < empleados->getLength(); i++){
        cout << empleados->getItem(i) << "\n";
        cout << "Sueldo total: " << empleados->getItem(i).calcularSueldo() << "\n";
    }
}

void Nomina::obtenerMaxMin(){
    Trabajador maxTrabajador = empleados->getItem(0);
    Trabajador minTrabajador = empleados->getItem(0);
    for(int i = 0; i < empleados->getLength(); i++) {
        if(empleados->getItem(i).getAnioIngreso() > maxTrabajador.getAnioIngreso()) {
            maxTrabajador = empleados->getItem(i);
        }
        if(empleados->getItem(i).getAnioIngreso() < minTrabajador.getAnioIngreso()) {
            minTrabajador = empleados->getItem(i);
        }
    }

    cout << "Trabajadores con mayor y menor antiguedad en : " << nombreCompania << "\n";
    cout << "Primer trabajador encontrado con la menor(-) antiguedad: " << "\n";
    cout << minTrabajador << "\n";
    cout << "Primer trabajador encontrado con la mayor(+) antiguedad: " << "\n";
    cout << maxTrabajador << "\n";
}

File: 03-Empresa/python/clase_nomina.py
from asyncio.windows_events import NULL
from clase_arreglo import Arreglo
from clase_trabajador import Trabajador

class Nomina:
    SEPARADOR_SALTO_DE_LINEA = "\n"
    SEPARADOR_COMA = ","
    empleados = NULL;
    nombreCompania = ""
    rutaArchivo = ""
    
    def __init__(self, rutaArchivo = "", nombreCompania = ""):
        self.rutaArchivo = rutaArchivo
        self.nombreCompania = nombreCompania
        content = ""

        entrada = open(self.rutaArchivo, "r")
        entrada.readline()
        linea = entrada.readline()
        while linea != "":
            content += linea
            linea = entrada.readline()

        separadoPorSaltosDeLinea = content.split(self.SEPARADOR_SALTO_DE_LINEA)

        self.empleados = Arreglo(len(separadoPorSaltosDeLinea))
        for i in range(len(separadoPorSaltosDeLinea)):
            separadoPorComas = separadoPorSaltosDeLinea[i].split(self.SEPARADOR_COMA)
            trabajadorTemp = Trabajador()
            trabajadorTemp.setNumeroTrabajador(int(separadoPorComas[0]))
            trabajadorTemp.setNombre(separadoPorComas[1])
            trabajadorTemp.setPaterno(separadoPorComas[2])
            trabajadorTemp.setMaterno(separadoPorComas[3])
            trabajadorTemp.setHorasExtra(int(separadoPorComas[4]))
            trabajadorTemp.setSueldoBase(float(separadoPorComas[5]))
            trabajadorTemp.setAnioIngreso(int(separadoPorComas[6]))

            self.empleados.setItem(i, trabajadorTemp)

    def calcularSueldo(self):
        for i in range(self.empleados.getLength()):
            print(self.empleados.getItem(i));
            print("Sueldo total: " + str(self.empleados.getItem(i).calcularSueldo()));

    def obtenerMaxMin(self):
        maxTrabajador = self.empleados.getItem(0);
        minTrabajador = self.empleados.getItem(0);
        for currTrabajador in self.empleados: 
            if(currTrabajador.getAnioIngreso() > maxTrabajador.getAnioIngreso()) :
                maxTrabajador = currTrabajador
            if(currTrabajador.getAnioIngreso() < minTrabajador.getAnioIngreso()) :
                minTrabajador = currTrabajador

        print("Trabajadores con mayor y menor antiguedad en : " + self.nombreCompania);
        print("Primer trabajador encontrado con la menor(-) antiguedad: ");
        print(minTrabajador);
        print("Primer trabajador encontrado con la mayor(+) antiguedad: ");
        print(maxTrabajador);